jax/_src/lax/parallel.py

```diff
@@ -1094,6 +1182,78 @@ def _raise_to_shaped_abstract_eval(x, *, axis_name, **params):
 batching.fancy_primitive_batchers[ppermute_p] = _ppermute_batcher
 batching.skippable_batchers[ppermute_p] = partial(_names_in_param, 'axis_name')
 
+
+def _psend_lowering(ctx, x, *, axis_name, perm):
+  full_perm, other_args = _pcollectives_lowering_common(
+      ctx, axis_name=axis_name, perm=perm, op_name="psend"
+  )
+  token = hlo.create_token()
+  send_op = hlo.SendOp(
+      [x],
+      token,
+      source_target_pairs=mlir.dense_int_elements(full_perm),
+      **other_args,
+  )
+  sharding = xc.OpSharding()
```

Comment:
Why set a sharding? This seems weird

Reply:
`psend/precv` JAX APIs lower to `send/recv` HLO instructions, which are side-effecting and must have sharding. See `spmd_partitioner.cc:5613`

Reply:
Discussed offline to keep it as is for now and fixing it later since this is a first cut of the API.